src/EditorFeatures/Core/LanguageServer/EditorConfig/Features/Diagnostics/DiagnosticHandler.cs

```diff
@@ -0,0 +1,314 @@
+                result.Add(DiagnosticTag.Unnecessary);
+
+            if (diagnosticData.CustomTags.Contains(WellKnownDiagnosticTags.EditAndContinue))
+                result.Add(VSDiagnosticTags.EditAndContinueError);
+
+            return result.ToArray();
+        }
+
+        protected override ValueTask<ImmutableArray<IDiagnosticSource>> GetOrderedDiagnosticSourcesAsync(RequestContext context, CancellationToken cancellationToken)
+        {
+            return ValueTaskFactory.FromResult(GetRequestedDocument(context));
+        }
+
+        protected override VSInternalDiagnosticReport[]? CreateReturn(BufferedProgress<VSInternalDiagnosticReport> progress)
+        {
+            return progress.GetValues();
+        }
+
+        internal static ImmutableArray<IDiagnosticSource> GetRequestedDocument(RequestContext context)
+        {
+            if (context.AdditionalDocument == null)
+            {
+                return ImmutableArray<IDiagnosticSource>.Empty;
+            }
+
+            if (!context.IsTracking(context.AdditionalDocument.GetURI()))
+            {
+                return ImmutableArray<IDiagnosticSource>.Empty;
+            }
+
+            return ImmutableArray.Create<IDiagnosticSource>(new DocumentDiagnosticSource(context.AdditionalDocument));
+        }
+
+        private record struct DocumentDiagnosticSource(TextDocument Document) : IDiagnosticSource
+        {
+            public ProjectOrDocumentId GetId() => new(Document.Id);
+
+            public Project GetProject() => Document.Project;
+
+            public Uri GetUri() => Document.GetURI();
+
+            public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(IDiagnosticAnalyzerService diagnosticAnalyzerService, RequestContext context, DiagnosticMode diagnosticMode, CancellationToken cancellationToken)
+            {
+                var document = context.AdditionalDocument;
+                Contract.ThrowIfNull(document);
+
+                var workspace = document.Project.Solution.Workspace;
+
+                var filePath = document.FilePath;
+                Contract.ThrowIfNull(filePath);
+
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                var diagnostics = new List<DiagnosticData>();
+
+                var definedSettings = new HashSet<string>();
+
+                foreach (var line in text.Lines)
+                {
+                    var lineText = line.ToString().Trim();
+
+                    // If line only contains a comment ignore
+                    // # A comment
+                    if (lineText.StartsWith("#"))
+                    {
+                        continue;
+                    }
+
+                    // Check if the line contains a setting definition and a comment
+                    // setting_name = setting_value # A comment
+                    if (lineText.Contains('#'))
```

Comment:
I wonder if we could use this at all - https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/EditorConfigParser.cs,43

We'd write our own accumulator that stores the found values and if there are any duplicates per section
(example accumulator) - https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/NamingStyleOptionAccumulator.cs,25

THen we don't need to handle any of this parsing of comments, keys , etc on our own